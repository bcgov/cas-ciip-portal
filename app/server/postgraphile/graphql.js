const {
  createPostGraphileSchema,
  withPostGraphileContext,
} = require("postgraphile");
const { graphql } = require("graphql");
const authenticationPgSettings = require("./authenticationPgSettings");
const { pgPool, getDatabaseUrl } = require("../db/databaseConnectionService");
const postgraphileOptions = require("./postgraphileOptions");

const pgSettings = (req) => {
  const opts = {
    ...authenticationPgSettings(req),
  };
  return opts;
};

let postgraphileSchemaSingleton;

const postgraphileSchema = async () => {
  if (!postgraphileSchemaSingleton) {
    postgraphileSchemaSingleton = await createPostGraphileSchema(
      getDatabaseUrl(),
      process.env.DATABASE_SCHEMA || "ggircs_portal",
      postgraphileOptions()
    );
  }

  return postgraphileSchemaSingleton;
};

async function performQuery(query, variables, request) {
  const settings = pgSettings(request);

  return withPostGraphileContext(
    {
      pgPool,
      pgSettings: settings,
    },
    async (context) => {
      // `context` object, which should NOT be used outside of this
      // function.
      const result = graphql(
        await postgraphileSchema(),
        query,
        null,
        { ...context }, // You can add more to context if you like
        variables
      );
      return result;
    }
  );
}

module.exports = {
  performQuery,
};
