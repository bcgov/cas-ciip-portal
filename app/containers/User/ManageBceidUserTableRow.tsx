import React from "react";
import { Button } from "react-bootstrap";
import { graphql, createFragmentContainer, RelayProp } from "react-relay";
import { ManageBceidUserTableRow_user } from "ManageBceidUserTableRow_user.graphql";
import updateUserMutation from "mutations/user/updateUserMutation";
import Link from "next/link";

interface Props {
  relay: RelayProp;
  user: ManageBceidUserTableRow_user;
  key: string;
}
export const ManageBceidUserTableRowComponent: React.FunctionComponent<Props> = (
  props
) => {
  const { user } = props;

  const handleChange = async () => {
    const variables = {
      input: {
        id: user.id,
        ciipUserPatch: {
          allowUuidUpdate: !user.allowUuidUpdate,
        },
      },
    };
    await updateUserMutation(props.relay.environment, variables);
  };

  const url = `/analyst/organisation-requests?filterArgs=${encodeURIComponent(
    `{"user_id":${user.rowId}}`
  )}`;

  return (
    <tr>
      <td>{user.firstName}</td>
      <td>{user.lastName}</td>
      <td>{user.emailAddress}</td>
      <td>
        <Link href={`${url}`} passHref>
          <a
            href="#"
            target="_blank"
            rel="noopener noreferrer"
            className="alert-link"
          >
            Access Requests
          </a>
        </Link>
      </td>
      <td>
        {user.allowUuidUpdate ? (
          <Button
            variant="outline-primary"
            onClick={async () => handleChange()}
          >
            Disable New BCeID Login
          </Button>
        ) : (
          <Button variant="primary" onClick={handleChange}>
            Enable New BCeID Login
          </Button>
        )}
      </td>
    </tr>
  );
};

export default createFragmentContainer(ManageBceidUserTableRowComponent, {
  user: graphql`
    fragment ManageBceidUserTableRow_user on CiipUser {
      id
      rowId
      firstName
      lastName
      emailAddress
      allowUuidUpdate
    }
  `,
});
