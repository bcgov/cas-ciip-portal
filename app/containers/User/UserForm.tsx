import React, {useState} from 'react';
import {createFragmentContainer, graphql, RelayProp} from 'react-relay';
import JsonSchemaForm, {IChangeEvent} from 'react-jsonschema-form';
import {JSONSchema6} from 'json-schema';
import {UserForm_user} from 'UserForm_user.graphql';
import createUserMutation from '../../mutations/user/createUserMutation';
import updateUserMutation from '../../mutations/user/updateUserMutation';
import FormArrayFieldTemplate from '../Forms/FormArrayFieldTemplate';
import FormFieldTemplate from '../Forms/FormFieldTemplate';
import FormObjectFieldTemplate from '../Forms/FormObjectFieldTemplate';

interface Props {
  user?: UserForm_user;
  relay: RelayProp;
  uuid?: string;
  defaultGivenName?: string;
  defaultFamilyName?: string;
  defaultEmail?: string;
  onSubmit: () => any;
}

const UserForm: React.FunctionComponent<Props> = ({
  user,
  uuid,
  defaultGivenName,
  defaultFamilyName,
  defaultEmail,
  relay,
  onSubmit
}) => {
  const handleChange = async (e: IChangeEvent) => {
    if (user)
      await updateUserMutation(relay.environment, {
        input: {
          id: user.id,
          ciipUserPatch: {
            ...e.formData
          }
        }
      });
  };

  const handleSubmit = async (e: IChangeEvent) => {
    if (user) await handleChange(e);
    else
      await createUserMutation(relay.environment, {
        ...e.formData,
        uuid
      });
    onSubmit();
  };

  const [userSchema] = useState<JSONSchema6>({
    type: 'object',
    properties: {
      firstName: {
        type: 'string',
        title: 'First Name',
        default: defaultGivenName
      },
      lastName: {
        type: 'string',
        title: 'Last Name',
        default: defaultFamilyName
      },
      emailAddress: {
        type: 'string',
        title: 'Email Address',
        default: defaultEmail
      },
      phoneNumber: {
        type: 'string',
        title: 'Phone Number'
      },
      occupation: {
        type: 'string',
        title: 'Occupation'
      }
    },
    required: [
      'firstName',
      'lastName',
      'emailAddress',
      'phoneNumber',
      'occupation'
    ]
  });

  return (
    <JsonSchemaForm
      omitExtraData
      liveOmit
      schema={userSchema}
      formData={user}
      showErrorList={false}
      ArrayFieldTemplate={FormArrayFieldTemplate}
      FieldTemplate={FormFieldTemplate}
      ObjectFieldTemplate={FormObjectFieldTemplate}
      onChange={handleChange}
      onSubmit={handleSubmit}
    />
  );
};

export {UserForm as UserFormComponent};
export default createFragmentContainer(UserForm, {
  user: graphql`
    fragment UserForm_user on CiipUser {
      id
      firstName
      lastName
      emailAddress
      phoneNumber
      occupation
    }
  `
});
