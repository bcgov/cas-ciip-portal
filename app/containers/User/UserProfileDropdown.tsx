import React from "react";
import { graphql, createFragmentContainer } from "react-relay";
import { Form, Dropdown } from "react-bootstrap";
import Link from "next/link";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faUser } from "@fortawesome/free-solid-svg-icons";
import { UserProfileDropdown_user } from "__generated__/UserProfileDropdown_user.graphql";
import { ButtonCSS } from "components/Layout/Header";

interface Props {
  user: UserProfileDropdown_user;
}

// Bootstrap's Dropdown.Item component doesn't naturally tab and submit the Form item.
const submitForm = (e) => {
  if (e.key === "Enter" || e.which === 13) {
    e.currentTarget.submit();
  }
};

const UserProfileDropdown: React.FunctionComponent<Props> = ({ user }) => {
  return (
    <>
      <li className="d-none d-lg-block">
        <Dropdown alignRight style={{ paddingLeft: "6px" }}>
          <Dropdown.Toggle
            id="user-icon"
            as="button"
            variant="primary"
            aria-label="User menu toggle"
            style={{
              borderRadius: "50%",
              width: 40,
              height: 40,
              border: "none",
              background: "#fff",
              position: "relative",
            }}
          >
            <FontAwesomeIcon
              color="#036"
              icon={faUser}
              style={{
                fontSize: "30px",
                verticalAlign: "middle",
                height: "0.8em",
                paddingBottom: "2px",
              }}
            />
          </Dropdown.Toggle>
          <Dropdown.Menu>
            <Dropdown.Item as={Link} href="/user/profile">
              <a className="dropdown-item text-right">
                <div>
                  <span>{user.firstName}</span> <span>{user.lastName}</span>
                </div>
                <div className="small text-muted">{user.emailAddress}</div>
              </a>
            </Dropdown.Item>
            <Dropdown.Item
              as={Form}
              action="/logout"
              method="post"
              tabIndex={0}
              onKeyPress={submitForm}
            >
              <button type="submit" className="w-100 text-right">
                Logout
              </button>
            </Dropdown.Item>
          </Dropdown.Menu>
        </Dropdown>
      </li>
      <li className="d-lg-none">
        <Link href="/user/profile">
          <a className="nav-button text-right">
            <div>
              <span>{user.firstName}</span> <span>{user.lastName}</span>
            </div>
            <div className="small">{user.emailAddress}</div>
          </a>
        </Link>
      </li>
      <li className="d-lg-none">
        <Form action="/logout" method="post">
          <button type="submit" className="nav-button text-right">
            Logout
          </button>
        </Form>
      </li>
      <style jsx>{ButtonCSS}</style>
      <style jsx global>{`
        #user-icon.dropdown-toggle::after {
          position: absolute;
          right: -14px;
          top: 46%;
          color: #fff;
        }
      `}</style>
    </>
  );
};

export { UserProfileDropdown as UserProfileDropdownComponent };
export default createFragmentContainer(UserProfileDropdown, {
  user: graphql`
    fragment UserProfileDropdown_user on CiipUser {
      firstName
      lastName
      emailAddress
    }
  `,
});
