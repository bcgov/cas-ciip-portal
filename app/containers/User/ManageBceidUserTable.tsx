import React from "react";
import { graphql, createFragmentContainer } from "react-relay";
import { ManageBceidUserTable_query } from "ManageBceidUserTable_query.graphql";
import ManageBceidUserTableRow from "./ManageBceidUserTableRow";
import {
  TableFilter,
  TextFilter,
  DisplayOnlyFilter,
  NoHeaderFilter,
} from "components/FilterableTable/Filters";
import FilterableTable from "components/FilterableTable/FilterableTable";

interface Props {
  query: ManageBceidUserTable_query;
}

const filters: TableFilter[] = [
  new TextFilter("First Name", "first_name"),
  new TextFilter("Last Name", "last_name"),
  new TextFilter("Email", "email_address"),
  new DisplayOnlyFilter("View Access Requests"),
  new NoHeaderFilter(),
];

export const ManageBceidUserTableComponent: React.FunctionComponent<Props> = (
  props
) => {
  const { edges, totalCount } = props.query.allCiipUsers;

  const body = (
    <tbody>
      {edges.map((edge) => (
        <ManageBceidUserTableRow key={edge.node.id} user={edge.node} />
      ))}
    </tbody>
  );
  return (
    <FilterableTable
      body={body}
      filters={filters}
      paginated
      totalCount={totalCount}
    />
  );
};

export default createFragmentContainer(ManageBceidUserTableComponent, {
  query: graphql`
    fragment ManageBceidUserTable_query on Query
    @argumentDefinitions(
      first_name: { type: "String" }
      last_name: { type: "String" }
      email_address: { type: "String" }
      order_by: { type: "[CiipUsersOrderBy!]" }
      pageSize: { type: "Int" }
      offset: { type: "Int" }
    ) {
      allCiipUsers(
        first: $pageSize
        offset: $offset
        filter: {
          uuid: { notIncludesInsensitive: "idir" }
          firstName: { includesInsensitive: $first_name }
          lastName: { includesInsensitive: $last_name }
          emailAddress: { includesInsensitive: $email_address }
        }
        orderBy: $order_by
      ) {
        edges {
          node {
            id
            ...ManageBceidUserTableRow_user
          }
        }
        totalCount
      }
    }
  `,
});
