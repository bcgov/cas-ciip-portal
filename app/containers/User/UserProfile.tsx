import React, { useState } from "react";
import { Button, Table } from "react-bootstrap";
import { graphql, createFragmentContainer } from "react-relay";
import UserForm from "./UserForm";

const edit = {
  width: "75px",
};
const table = {
  marginTop: "30px",
  marginBottom: "50px",
};

const UserProfile = ({ user }) => {
  const [editMode, setEditMode] = useState(false);
  return (
    <div>
      <Button
        className="button-edit"
        style={edit}
        variant="primary"
        onClick={() => setEditMode(true)}
      >
        Edit
      </Button>
      <p />
      {editMode && <UserForm user={user} onSubmit={() => setEditMode(false)} />}
      {!editMode && (
        <Table striped bordered hover style={table}>
          <tbody>
            <tr>
              <th>First Name</th>
              <td>{user.firstName}</td>
            </tr>
            <tr>
              <th>Last Name</th>
              <td>{user.lastName}</td>
            </tr>
            <tr>
              <th>Occupation</th>
              <td>{user.occupation}</td>
            </tr>
            <tr>
              <th>Phone Number</th>
              <td>{user.phoneNumber}</td>
            </tr>
            <tr>
              <th>Email</th>
              <td>{user.emailAddress}</td>
            </tr>
          </tbody>
        </Table>
      )}
    </div>
  );
};

export { UserProfile as UserProfileComponent };
export default createFragmentContainer(UserProfile, {
  user: graphql`
    fragment UserProfile_user on CiipUser {
      ...UserForm_user
      firstName
      lastName
      emailAddress
      phoneNumber
      occupation
    }
  `,
});
