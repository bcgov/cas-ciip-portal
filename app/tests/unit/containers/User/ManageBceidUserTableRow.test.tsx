import React from "react";
import { shallow } from "enzyme";
import { ManageBceidUserTableRow_userOrganisation } from "__generated__/ManageBceidUserTableRow_userOrganisation.graphql";
import { ManageBceidUserTableRowComponent } from "containers/User/ManageBceidUserTableRow";

describe("Organisation request rows", () => {
  it("should render an organisation properly", async () => {
    const user: ManageBceidUserTableRow_userOrganisation = {
      " $refType": "ManageBceidUserTableRow_userOrganisation",
      emailAddress: "a@a.a",
      firstName: "First",
      lastName: "Last",
      allowUuidUpdate: "false",
      id: "abc",
      rowId: 1,
    };

    const r = shallow(
      <ManageBceidUserTableRowComponent user={user} key="1" relay={null} />
    );
    expect(r).toMatchSnapshot();
  });
});
