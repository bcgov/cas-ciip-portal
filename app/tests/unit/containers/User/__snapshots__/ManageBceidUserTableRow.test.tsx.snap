// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organisation request rows should render an organisation properly 1`] = `
<tr>
  <td>
    First
  </td>
  <td>
    Last
  </td>
  <td>
    a@a.a
  </td>
  <td>
    <Link
      href="/analyst/organisation-requests?filterArgs=%7B%22user_id%22%3A1%7D"
      passHref={true}
    >
      <a
        className="alert-link"
        href="#"
        rel="noopener noreferrer"
        target="_blank"
      >
        Access Requests
      </a>
    </Link>
  </td>
  <td>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="outline-primary"
    >
      Disable New BCeID Login
    </Button>
  </td>
</tr>
`;
